src/persist-client/src/iter.rs

```diff
@@ -621,8 +632,18 @@ where
         let mut ready_futures: FuturesUnordered<_> = self.runs[0..first_larger]
             .iter_mut()
             .map(|run| async {
-                let part = &mut run.front_mut().expect("trimmed run should be nonempty").0;
-                if let ConsolidationPart::Queued { data, task } = part {
+                // It's possible for there to be multiple layers of indirection between us and the first available encoded part:
+                // if the first part is a `HollowRuns`, we'll need to fetch both that and the first part in the run to have data
+                // to consolidate. So: we loop, and bail out of the loop when either the first part in the run is available or we
+                // hit some unrecoverable error.
+                'fetch_loop: loop {
```

Comment:
The control flow here is a bit hard to follow (also I usually find naming loops to be a code smell). Is it possible to pull this out into a method? (Fine if the answer is "no" or "not worth it")

Reply:
I did try this and got stuck, so I'm hesitant to try again. 😅

I can swap those labelled breaks for returns, though, if you prefer it!

Reply:
(I did the swap, since it seems at least as clear and a bit more normal...)